### PR TITLE
PCHR-1275: PublicHoliday API,Model,ModelInstance

### PR DESCRIPTION
**Pull request by dhavalsharma**
# PublicHoliday 
Implementation for publicholiday api end points as per ticket [PCHR-1275](https://compucorp.atlassian.net/browse/PCHR-1275).

## API
`PublicHolidayAPI` will get all the publicholidays from backends.

## Model
`PublicHoliday` contains the publicholiday data and is wrapper over the API. It also has method `isPublicHoliday` to check if given date is public holiday or not.

## Instance 
`PublicHolidayInstance` is the publicholiday instance with helper method to get publicholiday data.

## Source Code location
It is located under `civihr/uk.co.compucorp.civicrm.hrleaveandabsences/js/angular`.

The implementation is located under shared folder. 

The test cases are in tests folder. 

## Sample PublicHoliday
A sample absence is presented below 

```json
{
  "is_error": 0,
  "version": 3,
  "count": 16,
  "values": [{
    "id": "1",
    "title": "New Year's Day",
    "date": "2016-01-01",
    "is_active": "1"
  }, {
    "id": "2",
    "title": "Good Friday",
    "date": "2016-03-25",
    "is_active": "1"
  }, {
    "id": "3",
    "title": "Easter Monday",
    "date": "2016-03-28",
    "is_active": "1"
  }, {
    "id": "4",
    "title": "Early May bank holiday",
    "date": "2016-05-02",
    "is_active": "1"
  }, {
    "id": "5",
    "title": "Spring bank holiday",
    "date": "2016-05-30",
    "is_active": "1"
  }, {
    "id": "6",
    "title": "Summer bank holiday",
    "date": "2016-08-29",
    "is_active": "1"
  }, {
    "id": "7",
    "title": "Boxing Day",
    "date": "2016-12-26",
    "is_active": "1"
  }, {
    "id": "8",
    "title": "Christmas Day (substitue day)",
    "date": "2016-12-27",
    "is_active": "1"
  }, {
    "id": "9",
    "title": "New Year's Day (substitute day)",
    "date": "2017-01-02",
    "is_active": "1"
  }, {
    "id": "10",
    "title": "Good Friday",
    "date": "2017-04-14",
    "is_active": "1"
  }, {
    "id": "11",
    "title": "Easter Monday",
    "date": "2017-04-17",
    "is_active": "1"
  }, {
    "id": "12",
    "title": "Early May bank holiday",
    "date": "2017-05-01",
    "is_active": "1"
  }, {
    "id": "13",
    "title": "Spring bank holiday",
    "date": "2017-05-29",
    "is_active": "1"
  }, {
    "id": "14",
    "title": "Summer bank holiday",
    "date": "2017-08-28",
    "is_active": "1"
  }, {
    "id": "15",
    "title": "Christmas day",
    "date": "2017-12-25",
    "is_active": "1"
  }, {
    "id": "16",
    "title": "Boxing day",
    "date": "2017-12-26",
    "is_active": "1"
  }]
}
```

### CODE REVIEW

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/apis/public-holiday-api.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+       * @return {Promise}
+       */
+      all: function (params) {
+        $log.debug('PublicHolidayAPI');
+
+        return this.sendGET('PublicHoliday', 'get', params)
+          .then(function (data) {
+            return data.values;
+          });
+      }
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/apis/public-holiday-api.js]: missing `;`

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/public-holiday-model.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+            .then(function (publicHolidays) {
+              return publicHolidays.map(function (publicHoliday) {
+                return instance.init(publicHoliday, true);
+              });
+            });
+        },
+        /**
+         *  Finds out if given date is a public holiday.
+         *
+         * @return {Bool} returns true if date is a public holday else false
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/public-holiday-model.js]: you are missing the documentation of `whichDate` parameter

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/public-holiday-model.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+         */
+        isPublicHoliday: function (whichDate) {
+          var dateFormat = HR_settings.DATE_FORMAT.toUpperCase();
+          var checkDate = moment(whichDate).format(dateFormat);
+
+          var params = {
+            'sequential': 1,
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/public-holiday-model.js]: isn't `sequential` always provided in the parent `api` service?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/public-holiday-model.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+            'date': checkDate
+          };
+
+          return publicHolidayAPI.all(params)
+            .then(function (publicHolidays) {
+              if (publicHolidays.length) {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/public-holiday-model.js]:
```js
return publicHolidays.length;
```
of
```js
return !!publicHolidays.length;
```
if you really want to enforce it to be boolean

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/models/public-holiday-model_test.js [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        beforeEach(function () {
+          promise = PublicHoliday.isPublicHoliday(testDate);
+        });
+
+        afterEach(function () {
+          //to excute the promise force an digest
+          $rootScope.$apply();
+        });
+
+        it('calls equivalent API method', function () {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/models/public-holiday-model_test.js]: you already tested for this

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/apis/public-holiday-api.js (hunk 1 above)
Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/apis/public-holiday-api.js]: thats a function end }

**Pull request merged**